// national-geographic-api/dist/util/Utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = __importDefault(require("request"));
const Configs_1 = require("../config/Configs");
class Utils {
    /* istanbul ignore next line */
    constructor() { }
    static makeRequest(options, cb) {
        if (typeof cb !== 'function') {
            return new Promise((resolve, reject) => {
                request_1.default(options, (err, resp, body) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(JSON.parse(body));
                    }
                });
            });
            // tslint:disable-next-line
        }
        else {
            request_1.default(options, (err, resp, body) => {
                if (err)
                    return cb(err);
                if (body) {
                    const result = JSON.parse(body);
                    if (result.error) {
                        return cb(result.error);
                        // tslint:disable-next-line
                    }
                    else {
                        return cb(null, result);
                    }
                }
                cb(null, resp);
            });
        }
    }
    static getRequestPayload(url) {
        return {
            url,
            method: 'GET',
            headers: {
                'apiauth-apiuser': Configs_1.Configs.getAPIUser(),
                'apiauth-apikey': Configs_1.Configs.getAPIKey(),
            },
        };
    }
    static getLatestNewsUrl(pageSize, pageNumber) {
        return `${Configs_1.Configs.getLatestNewsAPIUrl()}&pageSize=${pageSize}&page=${pageNumber}`;
    }
    static getPictureOfDayUrl(startDate, numberOfDays) {
        return `${Configs_1.Configs.getPictureOfDayAPIUrl()}&publication_datetime_from=${startDate}T00:00:00Z&page=${numberOfDays}&limit=1`;
    }
}
exports.Utils = Utils;
